HCore-Web/Providers/Impl/DownloadProcessingProxyUrlProviderImpl.cs

```diff
@@ -0,0 +1,197 @@
+﻿using System;
+using System.IO;
+using System.Net.Http;
+using System.Security.Cryptography;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.DataProtection;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json;
+using static System.Convert;
+using static Newtonsoft.Json.JsonConvert;
+
+namespace HCore.Web.Providers.Impl
+{
+    /// <summary>
+    /// Defines a service to facade downloads with a download proxy, which might perform some processing of downloads.
+    ///
+    /// <remarks>There is a need for some downloads from external sources to rename the download file name.
+    /// Unfortunately this is not possible with external URLs. Only a proxy facade may do so by
+    /// downloading the file data on behalf of the user and passing it with a changed name.</remarks>
+    ///
+    /// <remarks>More processing is possible, of course.</remarks>
+    /// </summary>
+    public class DownloadProcessingProxyUrlProviderImpl : IDownloadProcessingProxyUrlProvider
+    {
+        public static readonly string ProcessingProxyBaseUrlConfig = "DownloadProxy:BaseUrl";
+        public static readonly string UrlQueryKeyName = "data";
+
+        private readonly ILogger<DownloadProcessingProxyUrlProviderImpl> _logger;
+        private readonly IDataProtector _protector;
+        private readonly IServiceProvider _serviceProvider;
+
+        public DownloadProcessingProxyUrlProviderImpl(
+            IDataProtectionProvider protectionProvider,
+            IServiceProvider serviceProvider,
+            ILogger<DownloadProcessingProxyUrlProviderImpl> logger
+        )
+        {
+            if (protectionProvider == null)
+            {
+                throw new ArgumentNullException(nameof(protectionProvider));
+            }
+
+            _protector = protectionProvider.CreateProtector(nameof(DownloadProcessingProxyUrlProviderImpl));
+            _serviceProvider = serviceProvider;
+            _logger = logger;
+        }
+
+
+        public virtual Task<Uri> CreateProxyUrlAsync(Uri downloadSourceUri, string fileName, string proxyBaseUrl, string downloadMimeType = null)
+        {
+            if (downloadSourceUri == null)
+            {
+                throw new ArgumentNullException(nameof(downloadSourceUri));
+            }
+
+            if (!downloadSourceUri.IsAbsoluteUri)
+            {
+                throw new ArgumentException("Download source URI is not absolute URI!");
+            }
+
+            if (string.IsNullOrEmpty(proxyBaseUrl))
+            {
+                throw new ArgumentException("No proxy base URL has been provided!");
+            }
+
+            // convert the data to JSON string
+            string uriQueryPayload = ConvertParametersToQueryData(CalculateHashFromParameters(downloadSourceUri, fileName, downloadMimeType));
+
+            UriBuilder proxyUri = new UriBuilder(proxyBaseUrl);
```

Comment:
Uri processing can be done much more safe in ASP.NET Core

See QueryHelpers

E.g. in CLAPI-C ContentProviders\Impl\Providers\Getty\GettyPurchase.cs from line 30

Reply:
Thanks for the hint - this is much better!